Added MidiClip input and SelectFloatFromDict

HolgerFoerterer opened the pull request:
Implemented reading MIDI files (this creates a dictionary)
Implemented SelectFloatFromDict (to parse this dictionary for specific notes and controllers)

MidiClips are timeClips so you can change the position and speed of the clip.
Edit: Known bugs: Playing clips backwards is not optimized and may at the moment cause performance issues.

pixtur commented on T3/Gui/OutputUi/FloatDictOutputUi.cs:
```diff
@@ -0,0 +1,162 @@
+                if (v == null)
+                    return;
+
+                var count = Math.Min(1024, v.Count);
```
```suggestion
                var trackCount = v.Count.Clamp(0,1024);
```

HolgerFoerterer replied:
Perhaps `valueCount`? It was called `length` in FloatListOutputUi.cs which I found confusing as well.

pixtur commented on T3/Gui/OutputUi/FloatDictOutputUi.cs:
```diff
@@ -0,0 +1,162 @@
+                    }
+                }
+
+                var valueEnum = v.Values.GetEnumerator();
```
```suggestion
                var enumarator = v.Values.GetEnumerator();
```

But why using enumerator instead of `foreach(` ?

HolgerFoerterer replied:
...lack of C# experience from my side again.

pixtur commented on T3/Gui/OutputUi/FloatDictOutputUi.cs:
```diff
@@ -0,0 +1,162 @@
+        private static float _minFit = -1;
+        private static float _maxFit = 1;
+
+        private static List<List<float>> _previousValues;
```
I'm not sure if this should be static. We should probably reset this automatically if, when switching between slots.

HolgerFoerterer replied:
Should not really be static. Either we should create more than one [FloatDictOutputUi](https://github.com/still-scene/t3/pull/140/files/1df7cf7a51b14d0e81d7152012c2aa7ccba34bae#diff-c16f25973d28ba49326a85486dc79916770be7eb913ef0a0de6da3c5eeae8354) instance, keep track of the specific dictionaries or reset on switching, yes.

pixtur reviewed (commented):
Excellent job!

Most of my comments are simple style and variable naming issues that would be easy to fix.

My main questions are:

- Is a generic `Dict<T>` class justified?
- Should we split the [MidiClip] into to operators [ReadMidiFile] and [MidiClip] ? This would probably simplify the interface by allowing to pass a time attribute.
- Should we clarify the static value list in `FloatDictOutput.cs`?
- Should we introduce a preparsing step on load time to collect all used channels and allow for faster iterating over the midi events?

HolgerFoerterer commented:
> * Is a generic `Dict<T>` class justified?

I introduced it when I tried to cope with the creation of ops and did not know that I had to register the UIs per value type. It's not necessary. I can remove it.

> * Should we split the [MidiClip] into to operators [ReadMidiFile] and [MidiClip] ? This would probably simplify the interface by allowing to pass a time attribute.

This would affect the time interface alone, which I don't think is really over-complicated. Separating these two parts may also cause new trouble in handling the events since you potentially have to handle multiple events during one time step and send these events back and forth between ops.

> * Should we clarify the static value list in `FloatDictOutput.cs`?

Yes, would be great. Simply making it non-static does not work though since the system seems to generate only one view. And there is a problem with the mouse-over evaluation which leads to flickering values in case you output one clip and mouse-over another. Also, due to the thumbnail calculation, the value printing will speed up by x2 in comparison to the case where you don't mouse-over. I have at the moment no idea how to solve this best without changing the underlying system or introducing a dictionary of nested lists...

> * Should we introduce a preparsing step on load time to collect all used channels and allow for faster iterating over the midi events?

Good idea. I don't know how much faster it would be but it may facilitate parsing the values backwards. We should keep all events so invisible value updates would still create an output change.

HolgerFoerterer commented on Core/DataTypes/Dict.cs:
```diff
@@ -0,0 +1,91 @@
+            // TODO: unverified...
+            writer.WritePropertyName("Dict");
+            writer.WriteStartArray();
+            foreach (var kvp in this)
```
Good idea. Please keep in mind that I started programming C# one month ago and my style may be horrible at the moment.

pixtur replied:
Don't worry. Style is trivial. Coding is the hard part, and your code is great!

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+        public readonly TimeClipSlot<Command> TimeSlot = new();
+        #endregion
+
+        [Input(Guid = "71655B86-B0ED-422E-89E3-E678C87A7E0E")]
```
So if I understand you correctly, inputs should be at the very end of the .cs, not the end of the pins? I picked the style of putting them directly below the outputs up from the old PlayVideo.cs.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+                    SetupMidiFile(context);
+                }
+
+                if (_midiEventCollection == null) return;
```
Ok.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+        }
+
+        // from https://github.com/naudio/NAudio/blob/master/Docs/MidiFile.md
+        private static string ToMBT(long eventTime, int ticksPerQuarterNote, TimeSignatureEvent timeSignature)
```
"Measures Beats Ticks". I mainly copied this code from the naudio documentation at the end of the page https://github.com/naudio/NAudio/blob/master/Docs/MidiFile.md

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+        private bool _initialized = false;
+        private MF _midiFile = null;
+        private MidiEventCollection _midiEventCollection = null;
+        private int _deltaTicksPerQuarterNote = 500000 / 60;
```
This is the default length of a MIDI tick in microseconds (60 ticks per beat if a beat is 500000 microseconds - which equals 120bps). I think it should be a `double` if done correctly but I took over the `int` from the naudio caluclations. Please see https://www.recordingblogs.com/wiki/time-division-of-a-midi-file for this information.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+        private double _lastTimeInBars = 0f;
+        private List<int> _lastEventOfTrack = null;
+        private bool _printLogMessages = false;
+        private bool _valuesUpdated = false;
```
Yes, possible.

HolgerFoerterer commented on T3/Gui/OutputUi/FloatDictOutputUi.cs:
```diff
@@ -0,0 +1,162 @@
+            {
+                ImGui.BeginChild("##scrolling");
+
+                var v = typedSlot.Value;
```
This code was borrowed from FloatListOutputUi.cs. The dictionary itself may be tracks or something else. I would keep value here.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+
+        public MidiClip()
+        {
+            _initialized = false;
```
Good. Lack of C# experience on my side.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+        {
+            var strictMode = false;
+            var filename = Filename.GetValue(context);
+            if (filename != null && filename.Length != 0)
```
Both agreed.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+using T3.Core.Operator.Attributes;
+using T3.Core.Operator.Slots;
+using NAudio.Midi;
+using MF = NAudio.Midi.MidiFile;
```
I shortened it when our own operator was called "MidiFile" as well, not "MidiClip"... :)

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+        // parsing the file
+        private TimeSignatureEvent _timeSignature = null;
+        private double _lastTimeInBars = 0f;
+        private List<int> _lastEventOfTrack = null;
```
Ah. I usually name vectors as a singular in C++. If you use plural in C#, I think your first proposal would make a lot of sense.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+                var sourceRange = TimeSlot.TimeClip.SourceRange;
+
+                // get the time we should be at in the MIDI file according to the timeClip
+                var bars = context.LocalTime - timeRange.Start;
```
Sure.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+                if (bars >= minRange &&
+                    bars < minRange + Math.Abs(sourceRange.Duration))
+                {
+                    for (var n = 0; n < _midiFile.Tracks; n++)
```
trackIndex is good. The origin of the `n` lies again in the examples of naudio at https://github.com/naudio/NAudio/blob/master/Docs/MidiFile.md

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+            }
+        }
+
+        private void UpdateTrack(int track, double time)
```
Sure.

HolgerFoerterer commented on Operators/Types/lib/io/midi/MidiClip.cs:
```diff
@@ -0,0 +1,238 @@
+                                    _deltaTicksPerQuarterNote, _timeSignature));
+                }
+
+                if (events[nextEvent] is NoteOnEvent noteOnEvent)
```
Didn't know that you could cast it like that -> lack of C# experience...

pixtur commented:
> I introduced it when I tried to cope with the creation of ops and did not know that I had to register the UIs per value type. It's not necessary. I can remove it.
It's not urgent, though.

> This would affect the time interface alone, which I don't think is really over-complicated. Separating these two parts may also cause new trouble in handling the events since you potentially have to handle multiple events during one time step and send these events back and forth between ops.
I think I can give this a try. I'm not sure, how to reuse the parsing code, but it not that much...

> Yes, would be great
I can give it a try.
